src/plugins/discover/public/application/apps/main/services/use_saved_search.ts

```diff
@@ -0,0 +1,328 @@
+import { AutoRefreshDoneFn, search } from '../../../../../../data/public';
+import { calcFieldCounts } from '../utils/calc_field_counts';
+import { SEARCH_ON_PAGE_LOAD_SETTING } from '../../../../../common';
+import { validateTimeRange } from '../utils/validate_time_range';
+import { updateSearchSource } from '../utils/update_search_source';
+import { SortOrder } from '../../../../saved_searches/types';
+import { getDimensions, getChartAggConfigs } from '../utils';
+import { buildPointSeriesData, Chart } from '../components/chart/point_series';
+import { TimechartBucketInterval } from '../components/timechart_header/timechart_header';
+
+export interface UseSavedSearch {
+  refetch$: Subject<unknown>;
+  savedSearch$: SavedSearchSubject;
+  shouldSearchOnPageLoad: () => boolean;
+}
+
+export interface SavedSearchSubjectMessage {
+  bucketInterval?: TimechartBucketInterval;
+  chartData?: Chart;
+  fetchCounter?: number;
+  fetchError?: Error;
+  fieldCounts?: Record<string, number>;
+  hits?: number;
+  inspectorAdapters?: { requests: RequestAdapter };
+  rows?: ElasticSearchHit[];
+  state: string;
+}
+
+export type SavedSearchSubject = BehaviorSubject<SavedSearchSubjectMessage>;
+
+export const useSavedSearch = ({
+  indexPattern,
+  savedSearch,
+  searchSessionManager,
+  searchSource,
+  services,
+  state,
+  stateContainer,
+  useNewFieldsApi,
+}: {
+  indexPattern: IndexPattern;
+  savedSearch: SavedSearch;
+  searchSessionManager: DiscoverSearchSessionManager;
+  searchSource: SearchSource;
+  services: DiscoverServices;
+  state: AppState;
+  stateContainer: GetStateReturn;
+  useNewFieldsApi: boolean;
+}): UseSavedSearch => {
+  const { data, filterManager, uiSettings } = services;
+  const timefilter = data.query.timefilter.timefilter;
+
+  const shouldSearchOnPageLoad = useCallback(() => {
+    // A saved search is created on every page load, so we check the ID to see if we're loading a
+    // previously saved search or if it is just transient
+    return (
+      uiSettings.get(SEARCH_ON_PAGE_LOAD_SETTING) ||
+      savedSearch.id !== undefined ||
+      timefilter.getRefreshInterval().pause === false ||
+      searchSessionManager.hasSearchSessionIdInURL()
+    );
+  }, [uiSettings, savedSearch.id, searchSessionManager, timefilter]);
+
+  /**
+   * The observable the UI (aka React component) subscribes to get notified about
+   * the changes in the data fetching process (high level: fetching started, data was received)
+   */
+  const savedSearch$: SavedSearchSubject = useMemo(
+    () =>
+      new BehaviorSubject({
+        state: shouldSearchOnPageLoad() ? fetchStatuses.LOADING : fetchStatuses.UNINITIALIZED,
+      }),
+    [shouldSearchOnPageLoad]
+  );
+  /**
+   * The observable to trigger data fetching in UI
+   * By refetch$.next('reset') rows and fieldcounts are reset to allow e.g. editing of runtime fields
+   * to be processed correctly
+   */
+  const refetch$ = useMemo(() => new Subject(), []);
```

Comment:
for sure, gives useful additional information 👍

Reply:
We're using the `useMemo` here also kind of as a guarantee to only instantiate that subject once. Unfortunately `useMemo` explicitly does not give those guarantees:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.
> – from [React documentation](https://reactjs.org/docs/hooks-reference.html#usememo)

React usually recommend using `useRef` for the semantical storing of unchangable values. Since we have an "expensive" instantiation here (`new Subject()`, that we might not want to call every rendering, and `useRef` unfortunately doesn't allow for [lazy initialization](https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily), I'd recommend we maybe pull out the a bit cumbersome code into it's own hook that might be benefital across more code in Discover later on, something like (untested, so there might be some typing issues):

```ts
function useSingleton<T>(initalizor: () => T): T {
  const ref = useRef<T>();

  if (ref.current === null) {
    ref.current = initializor();
  }

  return ref.current;
}
```

Reply:
been there, done that 👍